### PR TITLE
add networkinterface spec with the ability to add multiple ip configurations for a vm

### PR DESCRIPTION
**Pull request by nijosmsft**
sdn cni allocates ip address for the pods from a sdn vnet. this provides a flat network for the pods and removes calico overlay. to support this, each node is allocated a predefined number of secondary ipconfigurations. 
this change adds a networkinterface spec to the azurestackhcimachines and azurestackhcivirtualmachines crds. cloudoperator will add the networkinterface spec with the count of secondary ipconfigurations. caph will use this config to create the secondary ips for the node in moc when the networkinterface resource is created.

### CODE REVIEW

File: api/v1alpha3/azurestackhcivirtualmachine_types.go [hunk 1]
```diff
@@ -49,6 +49,9 @@ type AzureStackHCIVirtualMachineSpec struct {
 	BackendPoolNames []string `json:"backendPoolNames,omitempty"`
 
 	AdditionalSSHKeys []string `json:"additionalSSHKeys,omitempty"`
+
+	// +optional
+	NetworkInterfaces NetworkInterfaces `json:"networkInterfaces,omitempty"`
```
Comment (zawachte) [inline api/v1alpha3/azurestackhcivirtualmachine_types.go:54]: lets just add this to the latest version of the api.

Note: We need to clean up these prior API versions.

File: config/default/manager_image_patch.yaml [hunk 2]
```diff
@@ -7,5 +7,5 @@ spec:
   template:
     spec:
       containers:
-      - image: mocimages.azurecr.io/caphcontroller:1.1.3
+      - image: mocimages.azurecr.io/caphcontroller:sdn-cni
```
Comment (zawachte) [inline config/default/manager_image_patch.yaml:10]: revert before checkin

File: api/v1beta1/types.go [hunk 3]
```diff
@@ -104,6 +104,41 @@ func (s Subnets) ToMap() map[string]*SubnetSpec {
...
+
+type IpConfigurationSpec struct {
+	Name string `json:"name,omitempty"`
+	// +optional
+	Primary bool `json:"primary,omitempty"`
+	// +optional
+	Allocation IPAllocationMethod `json:"allocation,omitempty"`
+	// below fields are unused, but adding for completeness
+	// +optional
+	Ipaddress string `json:"ipaddress,omitempty"`
```
Comment (zawachte) [inline api/v1beta1/types.go:123]:
```suggestion
	Ipaddress string `json:"ipAddress,omitempty"`
```

File: api/v1beta1/types.go [hunk 4]
```diff
@@ -104,6 +104,41 @@ func (s Subnets) ToMap() map[string]*SubnetSpec {
...
+	// +optional
+	Prefixlength string `json:"prefixlength,omitempty"`
```
Comment (zawachte) [inline api/v1beta1/types.go:125]:
```suggestion
	PrefixLength string `json:"prefixLength,omitempty"`
```

File: api/v1beta1/types.go [hunk 5]
```diff
@@ -104,6 +104,41 @@ func (s Subnets) ToMap() map[string]*SubnetSpec {
...
+	// +optional
+	Subnetid string `json:"subnetid,omitempty"`
```
Comment (zawachte) [inline api/v1beta1/types.go:127]:
```suggestion
	SubnetId string `json:"subnetId,omitempty"`
```

**Review by zawachte — DISMISSED** [review]

**Review by nwoodmsft — DISMISSED** [review]

**Review by zawachte — APPROVED** [review]

**Review by nwoodmsft — APPROVED** [review]

**Pull request merged**